### PR TITLE
refactor(dtype): move all the castable logic to a single function

### PR DESCRIPTION
**Pull request by kszucs**
This makes it easier to understand the implicit casting rules.

### CODE REVIEW

File: ibis/expr/datatypes/tests/test_cast.py [hunk 1]
```diff
@@ -1,8 +1,50 @@
 from __future__ import annotations
 
+import hypothesis as h
 import pytest
 
 import ibis.expr.datatypes as dt
+import ibis.tests.strategies as ibst
+
+
+@h.given(ibst.signed_integer_dtypes(), ibst.signed_integer_dtypes())
```
Comment (kszucs) [inline ibis/expr/datatypes/tests/test_cast.py:10]: I am planning to add more hypothesis test cases here.

File: ibis/expr/datatypes/cast.py [hunk 2]
```diff
@@ -24,16 +24,120 @@ def cast(source: str | dt.DataType, target: str | dt.DataType, **kwargs) -> dt.D
     return target
 
 
+@public
```
Comment (kszucs) [inline ibis/expr/datatypes/cast.py:27]: Should remove all of the public decorators in this module.

**Review by cpcloud — APPROVED** [review]

File: ibis/expr/datatypes/tests/test_cast.py (hunk 1 above)
Comment (kszucs) [reply ibis/expr/datatypes/tests/test_cast.py:10]: Going to address this in a follow-up.

File: ibis/expr/datatypes/cast.py (hunk 2 above)
Comment (kszucs) [reply ibis/expr/datatypes/cast.py:27]: Going to address this in a follow-up.

**Pull request merged**